src/io_uring.c

```diff
@@ -0,0 +1,91 @@
+#include "io_uring.h"
+
+#ifdef HAVE_LIBURING
+#include <liburing.h>
+#include <string.h>
+
+#include "zmalloc.h"
+
+#define IO_URING_DEPTH 256 /* io_uring instance queue depth. */
+
+struct io_uring_context {
+    struct io_uring ring;
+    size_t queue_len;
+};
+
+io_uring_context *createIOUring(void) {
+    io_uring_context *uring_context = zmalloc(sizeof(struct io_uring_context));
+    uring_context->queue_len = 0;
+    struct io_uring_params params;
+    memset(&params, 0, sizeof(params));
+    /* On success, io_uring_queue_init_params(3) returns 0 and ring will
+     * point to the shared memory containing the io_uring queues.
+     * On failure -errno is returned. */
+    if (io_uring_queue_init_params(IO_URING_DEPTH, &uring_context->ring, &params) < 0) return NULL;
+    return uring_context;
+}
+
+/* Submit fdatasync request to io_uring. */
+int ioUringPrepFsyncAndSubmit(io_uring_context *ring_context, int fd) {
+    struct io_uring_sqe *sqe = io_uring_get_sqe(&ring_context->ring);
+    io_uring_prep_fsync(sqe, fd, IORING_FSYNC_DATASYNC);
```

Comment:
@PingXie I used to check the sqe == NULL, but removed later due to 2 reasons:
1. The sqe == NULL happened when SQ ring is currently full and then we can submit the requests to make sqe allocated, in this function, it only prepared one fsync job and then submit together, and we defined io_uring queue depth 256,  it means we never happened the sqe == NULL.
2. Currently we return the errno to caller which is get from io_uring_submit.  If SQE == NULL really happened, which kind of errno should we return?

Reply:
In theory we could still exhaust the SQ, right? Imagine that for whatever reason, we "forgot" to call `ioUringWaitFsyncBarrier` 256 in a row or fsync got stuck 256 times, the next attempt to submit the fsync job will fail to get an sqe. Agreed that this is a very unlikely scenario if we factor in the use case (aof fsync) but I am looking strictly at the contract that `io_uring.c` is committed to. 

Error handling is indeed tricky now with our abstraction that hides away the concept of individual IO operations (SQE/CQE). This is why I started thinking if we should start returning `successfully completed count` in all places (https://github.com/valkey-io/valkey/pull/599#discussion_r1649582979) and have io_uring.c process `errno` and `serverLog` the detailed error messages. This would work as long as we don't need to propagate the detailed error message back to the `client`, in other words, if the errors are only a concern of admins/operators, which seems to be the case for io_uring.

Reply:
@PingXie , Maybe I didn't describe clearly to make you confused.

The sqe == null happened between `io_uring_prep_fsync` and `io_uring_submit`, imagine we call `io_uring_prep_fsync` 256 times but didn't call `io_uring_submit` once, this will cause sqe == null, but in our implementation, we call `io_uring_prep_fysnc` and then call `io_uring_submit` followed (1:1), the SQEs will be empty after `io_uring_submit`, it has  no relationship with CQEs (`ioUringWaitFsyncBarrier`).